[JSC] Implement `String.prototype.concat` in C++

sosukesuzuki opened the pull request:
#### 91036c3fd11247b926d7d43e2b228c2468935c5a
<pre>
[JSC] Implement `String.prototype.concat` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=303680">https://bugs.webkit.org/show_bug.cgi?id=303680</a>

Reviewed by Yusuke Suzuki.

This patch implements `String.prototype.concat` in C++ with the following changes:

- Adds a new use kind `NotOtherUse` to implement the `RequireObjectCoercible`
  abstract operation[1] using a `Check` node.
- Handles `String#concat` in DFG/FTL using the existing `StrCat` DFG node.
  `StrCat` is already used for template literals and well optimized.

[1]: <a href="https://tc39.es/ecma262/#sec-requireobjectcoercible">https://tc39.es/ecma262/#sec-requireobjectcoercible</a>

                                        TipOfTree                  Patched

string-prototype-concat-2-args        1.8105+-0.1984     ^      0.3626+-0.0556        ^ definitely 4.9929x faster
string-prototype-concat-5-args        2.7280+-0.0825     ^      0.4609+-0.0415        ^ definitely 5.9184x faster

* JSTests/microbenchmarks/string-prototype-concat-2-args.js:
* JSTests/microbenchmarks/string-prototype-concat-5-args.js: Added.
* JSTests/stress/string-prototype-concat-no-args.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat-this-null.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat-this-undefined.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat.js: Added.
(shouldBe):
(concat):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.stringConcatSlowPath): Deleted.
(concat): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::SafeToExecuteEdge::operator()):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGUseKind.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGUseKind.h:
(JSC::DFG::typeFilterFor):
(JSC::DFG::checkMayCrashIfInputIsEmpty):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304079@main">https://commits.webkit.org/304079@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b69742574521cf7d93610b076d6806829bef2cb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6975 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45745 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142044 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86486 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7578 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6839 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70088 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120555 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83608 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5157 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo-libwebrtc~~](https://ews-build.webkit.org/#/builders/166/builds/2660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126568 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38695 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144738 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133024 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6653 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39279 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6729 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5580 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111491 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28273 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4987 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116831 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60513 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6706 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35031 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165945 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70283 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43375 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6758 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6635 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3132,6 +3132,45 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+            }
+
+            constexpr unsigned maxStrCatArguments = 3;
+            Node* operands[AdjacencyList::Size];
```
You can just do ` = { };` to ensure that all of them are `nullptr`.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3132,6 +3132,45 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+            Node* operands[AdjacencyList::Size];
+            unsigned indexInOperands = 0;
+            for (unsigned i = 0; i < AdjacencyList::Size; ++i)
+                operands[i] = nullptr;
```
So you do not need this.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -3132,6 +3132,45 @@ auto ByteCodeParser::handleIntrinsicCall(Node* callee, Operand resultOperand, Ca
+            for (unsigned i = 0; i < AdjacencyList::Size; ++i)
+                operands[i] = nullptr;
+
+            operands[indexInOperands++] = thisNode;
```
Let's add a test `"string".concat()` which only has one node.

sosukesuzuki committed:
[JSC] Implement `String.prototype.concat` in C++
https://bugs.webkit.org/show_bug.cgi?id=303680

Reviewed by Yusuke Suzuki.

This patch implements `String.prototype.concat` in C++ with the following changes:

- Adds a new use kind `NotOtherUse` to implement the `RequireObjectCoercible`
  abstract operation[1] using a `Check` node.
- Handles `String#concat` in DFG/FTL using the existing `StrCat` DFG node.
  `StrCat` is already used for template literals and well optimized.

[1]: https://tc39.es/ecma262/#sec-requireobjectcoercible

                                        TipOfTree                  Patched

string-prototype-concat-2-args        1.8105+-0.1984     ^      0.3626+-0.0556        ^ definitely 4.9929x faster
string-prototype-concat-5-args        2.7280+-0.0825     ^      0.4609+-0.0415        ^ definitely 5.9184x faster

* JSTests/microbenchmarks/string-prototype-concat-2-args.js:
* JSTests/microbenchmarks/string-prototype-concat-5-args.js: Added.
* JSTests/stress/string-prototype-concat-no-args.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat-this-null.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat-this-undefined.js: Added.
(shouldBe):
(concat):
* JSTests/stress/string-prototype-concat.js: Added.
(shouldBe):
(concat):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.stringConcatSlowPath): Deleted.
(concat): Deleted.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::SafeToExecuteEdge::operator()):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGStrengthReductionPhase.cpp:
* Source/JavaScriptCore/dfg/DFGUseKind.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/dfg/DFGUseKind.h:
(JSC::DFG::typeFilterFor):
(JSC::DFG::checkMayCrashIfInputIsEmpty):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304079@main
